### PR TITLE
Manifest kubernetes infra images

### PR DESCRIPTION
**Pull request by mkumatag**
**What this PR does / why we need it**:
This PR is for creating manifest for the kubernetes infra images
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```
NONE
```

### CODE REVIEW

**Comment by mkumatag** [comment]
ping @ixdy @luxas

**Comment by mkumatag** [comment]
@ixdy @luxas can you please review this code?

File: cluster/images/etcd/Makefile
Comment (dims) [inline cluster/images/etcd/Makefile]: I was not expecting the version change here. Is there a reason we are doing that?

Comment (mkumatag) [reply cluster/images/etcd/Makefile]: lot of changes went in the latest etcd makefile, lemme rebase first.

**Comment by spiffxp** [comment]
/retest
needs rebase

**Comment by spiffxp** [comment]
/retest
needs rebase

File: build/debian-base/Makefile
Comment (dims) [reply build/debian-base/Makefile]: +1 to using `docker manifest`

Comment (tao12345666333) [reply build/debian-base/Makefile]: +1 to using `docker manifest`, although it is experimental.

**Comment by mkumatag** [comment]
Now, this PR is ready for review.. @ixdy @spiffxp @dims @tao12345666333 ptal

**Comment by mkumatag** [comment]
@neolit123

File: cluster/images/hyperkube/README.md
Comment (neolit123) [inline cluster/images/hyperkube/README.md]: tiny nit: extra space after `and`

**Review by neolit123 — COMMENTED** [review]
@mkumatag thanks, LGTM. added one tiny nit - can be ignored. :+1:

**Comment by spiffxp** [comment]
```
W0913 09:04:01.953] invalid argument "gcr.io/k8s-testimages/bootstrap:v20180912-f97e35a46-amd64:v1.13.0-alpha.0.1249_dc396299fe0e92" for "-t, --tag" flag: invalid reference format
```

File: build/debian-hyperkube-base/Makefile
Comment (ixdy) [reply build/debian-hyperkube-base/Makefile]: ... though apparently maybe we could use the `--platform` option with `docker pull` and `docker build`

Comment (ixdy) [reply build/debian-hyperkube-base/Makefile]: ... but that seems to require the server to be running in experimental mode. bleh.

File: build/debian-hyperkube-base/Makefile
Comment (ixdy) [inline build/debian-hyperkube-base/Makefile]: update this?

File: cluster/images/hyperkube/Makefile
Comment (ixdy) [inline cluster/images/hyperkube/Makefile]: I don't think you need to update the hyperkube Makefile; humans don't directly call `make` in here; instead, it's called through the build scripts `build/lib/release.sh`, and then your work in https://github.com/kubernetes/release/pull/622 ensures a manifest list is created for hyperkube.

**Comment by ixdy** [comment]
basically lgtm, though I don't think we need the hyperkube changes.

after this merges, we'll probably need someone to run just the `push-manifest` rules. if we instead want to fully rebuild all of the images, we'll need tag bumps as well.

**Comment by ixdy** [comment]
@spiffxp that failure is because we inject `IMAGE` into most of our CI runs, pointing at the kubekins image running that container. :scream:

**Comment by ixdy** [comment]
or I guess bootstrap here

**Comment by neolit123** [comment]
/priority critical-urgent
we need etcd manifest lists in 1.12.

everything else has them except for kube-dns and this.

**Comment by dims** [comment]
Do we need to get this into v1.12 @ixdy @mkumatag ?

( cc @tpepper @guineveresaenger )

**Comment by neolit123** [comment]
@dims 
we need the etcd manifest change at least for sure.

**Comment by spiffxp** [comment]
/milestone v1.12
SIG Testing chair, this helps with conformance test requirements to work across the same arches we build releases for

File: cluster/images/etcd/Makefile
Comment (dims) [inline cluster/images/etcd/Makefile]: Do we need to bump REVISION?

Comment (mkumatag) [reply cluster/images/etcd/Makefile]: do we really need to update the revision? @ixdy

Comment (dims) [reply cluster/images/etcd/Makefile]: we already have `3.2.24-0` in registry already @mkumatag ( https://console.cloud.google.com/gcr/images/google_containers/GLOBAL/etcd?gcrImageListsize=50 )

Comment (neolit123) [reply cluster/images/etcd/Makefile]: `3.2.24 != 3.2.24-0`

Comment (ixdy) [reply cluster/images/etcd/Makefile]: There's also a `3.2.24` already.

This might be problematic, since I don't think our image promotion process allows overwriting an existing tag.

**Comment by ixdy** [comment]
/cc @jpbetz 

(as owner of the etcd image)

**Comment by neolit123** [comment]
the `3.2.24` tag is already used in kubeadm and possibly in other places:
https://github.com/kubernetes/kubernetes/search?q=etcd+%223.2.24%22&unscoped_q=etcd+%223.2.24%22

i'm assuming that third parties might rely on the exact semantic `3.2.24` minus the `-0` as well?

**Comment by dims** [comment]
@ixdy as long as our promotion software is ok with overwrite, i am good!.

**Comment by ixdy** [comment]
viewed another way: are we 100% sure that everything that currently references `k8s.gcr.io/etcd:3.2.24` knows how to handle a manifest list? it makes me a bit nervous to change something that already exists.

**Comment by ixdy** [comment]
Glancing at that search, it looks like most things use `3.2.24-0` for the docker tag. Can we bump the docker tag to `3.2.24-1` and then update all references?

**Comment by ixdy** [comment]
(well, most things except kubeadm...)

**Comment by ixdy** [comment]
OK, so I think we can overwrite the existing tag without too much effort, but my concern in https://github.com/kubernetes/kubernetes/pull/59664#issuecomment-422922820 stands. Is anything using the existing tag in production right now? If so, I'm very loath to overwrite it.

**Comment by dims** [comment]
@ixdy we just added support for 3.2.24 last week, so we should be safe. https://github.com/kubernetes/kubernetes/pull/68318

File: cluster/images/etcd/Makefile
Comment (timothysc) [reply cluster/images/etcd/Makefile]: tags are mutable ;-) TBH this is googles release jiggery, so I don't have strong opinions wrt -revision.

**Comment by neolit123** [comment]
```
timothysc [< 1 minute ago]
tags are mutable and it’s more of a manifest issue then an image issue.

timothysc [< 1 minute ago]
Given that that version hasn’t been in the wild I think just mutate the tag.
```

@ixdy @dims 
just let me know what tag we end up with so that i can update kubeadm. :+1:

**Comment by ixdy** [comment]
OK, I appreciate the confirmation that `3.2.24` is new and basically unused. I think I'm OK with overwriting the existing tags in that case.

**Comment by dims** [comment]
cool. let's ship what's in here now

/lgtm

**Comment by jpbetz** [comment]
Yes `3.2.24` is new.  I'm okay with overwriting it with another image that has the etcd 3.2.24 binary in it. That's safe.

The quick background is: We introduced `<etcd-version>-<revision>` earlier this year to enable us to rev the image independent from the (highest) etcd version in the image, but for backward compatibility, we'll been continuing to publish the `<etcd-version>` tags.  I'd like to deprecate them, but that needs to be done carefully so that we don't break anything/anyone.

**Comment by jpbetz** [comment]
Just realized I didn't fully understand the manifest list replacement idea.  I'm fine with kicking off the manifest use list with 3.2.24 since we just published it recently.

File: cluster/images/etcd/Makefile
Comment (jpbetz) [reply cluster/images/etcd/Makefile]: Safest approach would be to bump REVISION to 1, publish 3.2.24-1, and overwrite 3.2.24 to also be the same image as 3.2.24-1.

We'll eventually deprecate the tags that don't have the revision suffix, but we're keeping it around for backward compatibility for now.

**Comment by spiffxp** [comment]
This LGTM as well but I don't have /approve rights for either of these dirs

**Comment by neolit123** [comment]
@dims 
do you want me to send a PR similar to this one that removes the `-0` suffix in certain places WRT to manifest list?
https://github.com/kubernetes/kubernetes/pull/68318

File: cluster/images/etcd/Makefile
Comment (ixdy) [reply cluster/images/etcd/Makefile]: I also like @jpbetz's idea ^. What do others think?

Comment (neolit123) [reply cluster/images/etcd/Makefile]: i think i'm OK with revisions as long as everything is manifest list. :)

**Comment by ixdy** [comment]
/hold cancel

I've pushed a commit to update the revision. PTAL?

File: cluster/images/etcd/Makefile
Comment (ixdy) [inline cluster/images/etcd/Makefile]: /hold

This looks wrong? https://github.com/etcd-io/etcd/blob/34fcabab55585b39835820278da4de65b3463358/CHANGELOG-3.2.md#v3224-2018-07-24

Comment (dims) [reply cluster/images/etcd/Makefile]: good catch, let's switch this back to 1.8.7 since that's the recommendation

Comment (ixdy) [reply cluster/images/etcd/Makefile]: /hold cancel

I pushed another commit reverting this

**Review by ixdy — APPROVED** [review]

**Comment by mkumatag** [comment]
Thanks @ixdy quick fix.. can someone give LGTM ?

**Comment by mkumatag** [comment]
Thanks all... Did anyone push the images? @ixdy @jpbetz

**Comment by ixdy** [comment]
I don't think we want to push any new images, just run `make push-manifest` using the existing images. (I guess maybe we need to rebuild `etcd` though, since its a different tag.)

I can look into this Thursday PT.

**Comment by ixdy** [comment]
Manifests and new etcd images have been pushed to staging-k8s.gcr.io. I'm working on promoting to k8s.gcr.io now.

I believe only kubeadm would be affected by this, since it uses the unrevisioned tag (`3.2.24`).
We probably want a follow-up PR to [update uses of `3.2.24-0`](https://cs.k8s.io/?q=3%5C.2%5C.24-0&i=nope&files=&repos=) to `3.2.24-1` as well as to use the manifest list `k8s.gcr.io/etcd` rather than the arch-specific suffixed images.

**Pull request merged**
